File: services/server/src/lottery/protocol/communication.py
import socket
from typing import List

import safe_socket
from lottery import Bet
from lottery.protocol.packets.header import (
    ACK_CODE,
    BET_CODE,
    END_CODE,
    HEADER_SIZE,
    Header,
)
from lottery.protocol.packets.packet import Packet

from services.server.src.lottery.protocol.packets.packet import MAX_BETS


def _send_packet(socket: socket.socket, packet: Packet) -> None:
    safe_socket.send_all(socket, packet.to_bytes())


def _recv_header(socket: socket.socket) -> Header:
    bytes = safe_socket.recv_all(socket, HEADER_SIZE)

    if not bytes:
        raise ConnectionError('connection closed while receiving header')

    header = Header.from_bytes(bytes)

    return header


def _expect_ack(socket: socket.socket, agency_id: int) -> None:
    header = _recv_header(socket)

    if header.code != ACK_CODE:
        raise ValueError('unexpected code from client')

    if header.agency_id != agency_id:
        raise ValueError('unexpected agency id')


def send_winners(socket: socket.socket, agency_id: int, winners: List[Bet]) -> None:
    offset = 0

    while offset < len(winners):
        bet_packet = Packet.new_bet_packet(agency_id, winners[offset:offset + MAX_BETS])
        _send_packet(socket, bet_packet)
        _expect_ack(socket, agency_id)

        offset += MAX_BETS

    end_packet = Packet.new_end_packet(agency_id)
    _send_packet(socket, end_packet)
    _expect_ack(socket, agency_id)


def recv_bets(socket: socket.socket) -> List[Bet]:
    header = _recv_header(socket)

    if header.code == END_CODE:
        _send_packet(socket, ack)
        return []

    if header.code != BET_CODE:
        raise ValueError('unexpected message from client')

    bytes = safe_socket.recv_all(socket, header.bet_size)

    if not bytes:
        raise ConnectionError('connection closed while receiving bets')

    bets = []
    Packet.add_bets_from_bytes(header, bytes, bets)

    _send_packet(socket, ack)

    return bets

File: services/client/src/lottery/protocol/communication.go
package protocol

import (
	"errors"
	"io"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/lottery"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/lottery/protocol/packets"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

func sendPacket(socket io.ReadWriter, packet *packets.Packet) error {
	bytes, err := packet.ToBytes()

	if err != nil {
		return err
	}
	err = safe_socket.SendAll(socket, bytes)

	if err != nil {
		return err
	}
	return nil
}

func recvHeader(socket io.ReadWriter, agencyId uint8) (*packets.Header, error) {
	bytes, err := safe_socket.RecvAll(socket, packets.HeaderSize)

	if err != nil {
		return nil, err
	}
	header, err := packets.HeaderFromBytes(bytes)

	if err != nil {
		return nil, err
	}
	if header.AgencyId != agencyId {
		return nil, errors.New("unexpected agency id")
	}
	return header, nil
}

func expectAck(socket io.ReadWriter, agencyId uint8) error {
	header, err := recvHeader(socket, agencyId)

	if err != nil {
		return err
	}
	if header.Code != packets.AckCode {
		return errors.New("unexpected code from server")
	}
	return nil
}

func SendBets(socket io.ReadWriter, agencyId uint8, bets []lottery.Bet) error {
	betPacket, err := packets.NewBetPacket(agencyId, bets)

	if err != nil {
		return err
	}
	err = sendPacket(socket, betPacket)

	if err != nil {
		return err
	}
	return expectAck(socket, agencyId)
}

func RecvWinners(socket io.ReadWriter, agencyId uint8) ([]lottery.Bet, error) {
	end := packets.NewEndPacket(agencyId)
	err := sendPacket(socket, end)

	if err != nil {
		return nil, err
	}
	err = expectAck(socket, agencyId)

	if err != nil {
		return nil, err
	}
	bets := make([]lottery.Bet, 0)
	ack := packets.NewAckPacket(agencyId)

	for {
		header, err := recvHeader(socket, agencyId)

		if err != nil {
			return nil, err
		}
		if header.Code == packets.EndCode {
			err = sendPacket(socket, ack)

			if err != nil {
				return nil, err
			}
			break
		}
		if header.Code != packets.BetCode {
			return nil, errors.New("unexpected message from server")
		}
		bytes, err := safe_socket.RecvAll(socket, int(header.BetSize))

		if err != nil {
			return nil, err
		}
		err = packets.AddBetsFromBytes(header, bytes, bets)

		if err != nil {
			return nil, err
		}
		err = sendPacket(socket, ack)

		if err != nil {
			return nil, err
		}
	}
	return bets, nil
}
